File: solutions/go/simple-linked-list/1/simple_linked_list.go
package linkedlist

import "errors"

type Node struct {
	data int
	next *Node
}

type List struct {
	head *Node
}

func New(elements []int) *List {
	var node *Node = nil
	for _, element := range elements {
		node = &Node{data: element, next: node}
	}
	return &List{head: node}
}

func (list *List) Size() (s int) {
	for node := list.head; node != nil; node = node.next {
		s++
	}
	return
}

func (list *List) Push(data int) {
	list.head = &Node{data: data, next: list.head}
}

func (list *List) Pop() (int, error) {
	if list.head == nil {
		return 0, errors.New("empty list")
	}
	data := list.head.data
	list.head = list.head.next
	return data, nil
}

func (list *List) Array() []int {
	size := list.Size()
	array := make([]int, size)
	i := size - 1
	for node := list.head; node != nil; node = node.next {
		array[i] = node.data
		i--
	}
	return array
}

func (list *List) Reverse() *List {
	out := New(nil)
	for node := list.head; node != nil; node = node.next {
		out.Push(node.data)
	}
	return out
}

File: solutions/cpp/simple-linked-list/1/simple_linked_list.cpp
#include "simple_linked_list.h"

#include <stdexcept>

namespace simple_linked_list {

using namespace std;

size_t List::size() const { return current_size; }

void List::push(int data) {
    head = new Element{data, head};
    current_size++;
}

int List::pop() {
    if (current_size == 0) {
        throw out_of_range("Cannot pop from an empty list");
    }

    auto data = head->data;
    auto old_head = head;
    head = old_head->next;
    delete old_head;
    current_size--;
    return data;
}

void List::reverse() {
    Element* prev = nullptr;
    Element* current = head;
    while (current != nullptr) {
        Element* next = current->next;
        current->next = prev;
        prev = current;
        current = next;
    }
    head = prev;
}

List::~List() {
    while (head != nullptr) {
        auto old_head = head;
        head = old_head->next;
        delete old_head;
    }
}

}  // namespace simple_linked_list
